### ISSUE TITLE
Sometime duplicate message?

### ISSUE THREAD

**Issue by bienhoang** [Schmavery/facebook-chat-api#465]
I use `api.sendMessage(body, message.threadID);` wih `body` is string, and some time message sended was duplicated.

**Comment by bobland1** [comment]
Yeah I know about this. It's due to when the user sends a message on a bad internet connection

**Comment by bobland1** [comment]
I am not sure why it happens. I did have a work around by checking the messageID before carrying on but that I longer works

**Comment by bobland1** [comment]
May change it from checking the messageID to the body

**Comment by bienhoang** [comment]
Ok. Thanks you.

**Comment by Schmavery** [comment]
@bobland1 how does your workaround work?  I don't understand.
@bienhoang is this on master or the version on npm?

**Comment by bobland1** [comment]
@Schmavery I run from the bleading edge and I am experiencing it. I have tested on both. My work around did work. But no doesnt as the messageID are different unlike before when each message unless duplicate was different. My workaround is as follows: 
`if (message.messageID != undefined) {
          var lastMsgCheck = fs.readFileSync("logs/messageCheck.log", "utf8");
          if (lastMsgCheck == message.messageID) {
              console.log("\x1b[32mcheck \x1b[35mmessageID\x1b[0mGot answer 404\n{ messageID: '"+lastMsgCheck+"',\n  check: true }")
              return true;
          } else {
              fs.writeFileSync("logs/messageCheck.log", message.messageID);
              console.log("\x1b[32mcheck \x1b[35mmessageID\x1b[0m Got answer 404\n{ messageID: '"+message.messageID+"',\n  check: false }")
          } ..... }`

**Comment by bobland1** [comment]
EDIT: It doesnt want to indent...

**Comment by ravkr** [comment]
@Schmavery I'm seeing duplicates about once per 2000-3000 messages :/
I suspect that it have something to do with network connection but I didn't have time to diagnose it

**Comment by Schmavery** [comment]
It seems we should either refine or disable the retry logic that was recently added?
Can anyone confirm that this line https://github.com/Schmavery/facebook-chat-api/blob/master/utils.js#L648 gets printed when we see the duplicate?

**Comment by ravkr** [comment]
I don't think it is related to any recent commit. I'm seeing it since the first time I have used facebook-chat-api :(

**Comment by GAMELASTER** [comment]
@Schmavery you are probably right, but that mean's the request returns a wrong result number and for this is resend done.

**Comment by bienhoang** [comment]
@Schmavery I'm using version ^1.3.0

**Comment by bobland1** [comment]
Is that from npm, or blessing edge

**Comment by bobland1** [comment]
I have experienced this on both

**Comment by alecmanabat** [comment]
I am experiencing this as well on npm version. It appears that the message goes through api.listen() twice because when i trigger a method that sends a random greeting as a response, two different greetings are sent.

**Comment by bobland1** [comment]
Are we able to have an update on this. From what I have found. Messages which take longer to send (bad internet connection) send two message receipts. And as the api is receiving both it replies to both.

The code I used:

if (message.messageID != undefined) {
      var lastMsgCheck = fs.readFileSync("logs/messageCheck.log", "utf8");
       if (lastMsgCheck == message.messageID) {
              console.log("\x1b[32mcheck \x1b[35mmessageID\x1b[0mGot answer 404\n{ messageID: '"+lastMsgCheck+"',\n  check: true }")
              return true;
       } else {
              fs.writeFileSync("logs/messageCheck.log", message.messageID);
              console.log("\x1b[32mcheck \x1b[35mmessageID\x1b[0m Got answer 404\n{ messageID: '"+message.messageID+"',\n  check: false }")
          } ..... }

Checks the messageID, as most of the time they have the same ID as it is the same message. But sometimes it isn't. That is the best solution from what I have found

**Comment by Schmavery** [comment]
@bobland1 I asked above for someone to provide some information from the logs, but didn't see anything regarding this, so it's hard to know how to proceed on a fix.  If it's not that, then I'm at somewhat of a loss for now.

> It seems we should either refine or disable the retry logic that was recently added?
Can anyone confirm that this line https://github.com/Schmavery/facebook-chat-api/blob/master/utils.js#L648 gets printed when we see the duplicate?

**Comment by Schmavery** [comment]
Actually, in this bug, does the message get sent twice or the callback just gets called twice?  It is not clear to me from this issue.

**Comment by bobland1** [comment]
The message gets sent once but the callback gets called twice I think. It's hard to get an example as you have to have a bad connection. When running the bot it shows it like there is 2 messages sent

**Comment by bobland1** [comment]
When the user only sent one, does that make sense? I'll be sure to send any examples if I get any

**Comment by Schmavery** [comment]
None of these code examples show how the callback is being handled.  What are the values of err and messageInfo for the 2 times the callback is called?

**Comment by bobland1** [comment]
The code I have shown, it checks per message the messageID and if it is the same as the previous messageID it will not carry on else it will

**Comment by bobland1** [comment]
There is not an error

**Comment by bsansouci** [comment]
@bobland1 I think @Schmavery's concern is whether this is due to FB's backends sending the message again, or if it's a bug in our logic which causes the callback to be called multiple times in certain configuration.

Has anyone experienced this recently?

**Comment by bobland1** [comment]
It is a completely random occurrence so it's hard to diagnose, my understanding of it must be that as due to the bad internet it sends off two messages receipts instead of one?

**Comment by ravkr** [comment]
few days ago I've had the same problem... with mobile connection (ping ~300ms instead of 10-20) I was getting sometimes even 6 messages with the same ID :cry: 

this is really weird bug ;(

**Comment by Schmavery** [comment]
Is anyone able to confirm that the message doesn't actually appear twice on facebook's end?
I wonder how we can filter out these duplicated without needing to remember state about previous messages...

**Comment by ravkr** [comment]
I can confirm it... 
when my bot runs on 100Mbps connection (ping 10-20ms), duplicated messages are almost nonexistent, but on mobile connection (512kbps, 300ms ping), almost all messages we duplicated (and these messages were shown in logs several times)

**Comment by ravkr** [comment]
there was nothing in between them

you can probably reproduce it by increasing ping...
e.g. use mobile connection, or change interface settings to use max 10Mb/s and then start downloading some Linux torrents
